iroh-io/src/lib.rs

```diff
@@ -0,0 +1,603 @@
+        }
+    }
+
+    fn async_test<F: Future>(f: F) -> F::Output {
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+        rt.block_on(f)
+    }
+
+    async fn write_op_test<W: AsyncSliceWriter, C: Fn(&W) -> Vec<u8>>(
+        ops: Vec<WriteOp>,
+        mut bytes: W,
+        content: C,
+    ) -> io::Result<()> {
+        let mut reference = Vec::new();
+        for op in ops {
+            apply_op(&mut reference, &op);
+            match op {
+                WriteOp::Write(offset, data) => {
+                    AsyncSliceWriter::write_at(&mut bytes, offset, data.into()).await?;
+                }
+                WriteOp::SetLen(offset) => {
+                    AsyncSliceWriter::set_len(&mut bytes, offset).await?;
+                }
+                WriteOp::Sync => {
+                    AsyncSliceWriter::sync(&mut bytes).await?;
+                }
+            }
+            assert_eq!(content(&bytes), reference.as_slice());
+        }
+        io::Result::Ok(())
+    }
+
+    async fn read_op_test<R: AsyncSliceReader>(
+        ops: Vec<ReadOp>,
+        mut file: R,
+        actual: &[u8],
+    ) -> io::Result<()> {
+        let mut current = 0u64;
+        for op in ops {
+            match op {
+                ReadOp::ReadAt(offset, len) => {
+                    let data = AsyncSliceReader::read_at(&mut file, offset, len).await?;
+                    current = offset.checked_add(len as u64).unwrap();
+                    assert_eq!(&data, &actual[limited_range(offset, len, actual.len())]);
+                }
+                ReadOp::ReadSequential(offset, len) => {
+                    let offset = if offset >= 0 {
+                        current.saturating_add(offset as u64)
+                    } else {
+                        current.saturating_sub((-offset) as u64)
+                    };
+                    let data = AsyncSliceReader::read_at(&mut file, offset, len).await?;
+                    current = offset.checked_add(len as u64).unwrap();
+                    assert_eq!(&data, &actual[limited_range(offset, len, actual.len())]);
+                }
+                ReadOp::Len => {
+                    let len = AsyncSliceReader::len(&mut file).await?;
+                    assert_eq!(len, actual.len() as u64);
+                }
+            }
+        }
+        io::Result::Ok(())
+    }
+
+    // #[cfg(feature = "http")]
+    // #[tokio::test]
+    // async fn test_http_range() -> io::Result<()> {
+    //     let url = reqwest::Url::parse("https://ipfs.io/ipfs/bafybeiaj2dgwpi6bsisyf4wq7yvj4lqvpbmlmztm35hqyqqjihybnden24/image").unwrap();
+    //     let mut resource = HttpAdapter::new(url).await?;
+    //     let buf = resource.read_at(0, 100).await?;
+    //     println!("buf: {:?}", buf);
+    //     let buf = resource.read_at(1000000, 100).await?;
+    //     println!("buf: {:?}", buf);
+    //     Ok(())
+    // }
```

Comment:
what's up with this test?

Reply:
It's just a thing to explore the behaviour of one particular http server that supports range requests. Since it is on IPFS I don't want to keep it enabled. I guess I could just disable it instead of commenting it out.